### ISSUE TITLE
Preprocessor in headers

### ISSUE THREAD

**Issue by Muzza** [slynch8/10x#554]
Very excited to see the pre-processor appearing. I use it a lot so you may get a lot of bug reports from me (sorry!).

The below is an example where it fails for me. When viewing header2.h it knows nothing of the FUNC define. I can understand that this happens because header2.h does not include header.h. But this kind of pattern appears everywhere, especially if using pre-compiled headers. 

```cpp
// header.h
#define FUNC( func ) func
```

```cpp
// header2.h
extern void FUNC(MyFunc)( void );
```

```cpp
// source.c
#include "header.h"
#include "header2.h"
```

**Comment by slynch8** [comment]
Hi, no problem. Interested to hear how you get on.

This is a something I hadn't thought about. You're right, it is quite common, but I'm not sure how to fix this. 10x parses files all files (header and cpp) individually and relies on things being included. I'm not sure when parsing a single header file how I can know what headers might be needed. Different cpp files could have different includes before they include the header. I'm open to ideas. I guess I could just search the whole project for a matching define and use that, but it won't always get it right and might cause other problems.

**Comment by slynch8** [comment]
I was thinking, I could go through all cpp files in the project that include the header and prepend anything included before the header to that header. Wouldn't be perfect but might be good enough for most cases.

I could also try and detect precompiled header usage but that might not catch everything.

**Comment by slynch8** [comment]
As a possible short term solution would it work if I give you a per project setting where you could specify includes that are always prepended to all files?

**Comment by Muzza** [comment]
Yes, a per-project 'global includes' setting would be a good short term measure.
As for parsing all cpp files, possibly it would be enough to just include the contents from any one cpp file that comes before the include? It has to be valid to compile. Including the contents from all cpp files could result in you parsing invalid code.

There are impossible cases too. Two cpp files can have intentionally different defines before including the header to get different behaviours. I suppose a really fancy feature would re-parse the header based upon the last cpp file opened that included it.

**Comment by slynch8** [comment]
I'm having a look at this now. Are you using PCH files? Would it be enough for me to detect any PCH files in the proj file and always include them?

**Comment by slynch8** [comment]
Alternatively I could just add the first include of all cpp files. Detecting the cpp lines previous to the include would be very complex to do.

**Comment by Muzza** [comment]
I use PCH in some projects but not others. I feel like using the includes from the first c/cpp file that uses the header would be the best solution. Being able to specify the associated c/cpp would be a nice advanced feature. Maybe a keyword within a comment in the header file could instruct 10x which c/cpp file to parse using.

**Comment by slynch8** [comment]
In the build I've just uploaded (0.29.22) I've added PCH support and the ability to set Force Includes in the workspace settings.

I agree that include from the first cpp that uses the header would be the ideal but it's not easy to get that information especially because it could be included indirectly, I think I'd have to add a whole new pass to the parser.

I'm tempted to just add the first include for all cpp files (if there is no PCH). A keyword is an interesting idea, I'll have a think about that.

**Comment by maxphilippov** [comment]
I'll bump this thread with just another use case. I was messing around with Handmade Hero source code. And force include doesn't solve this problem there since one header there defines:
```
#define global static
```
And another header doesn't include the first one (and force include doesn't work since it's a header-file) defines a variable:
```
global char *Functions;
```
And you can't neither go to definition of Functions nor find references.

**Comment by Muzza** [comment]
I should also confirm that whereas the implemented solution works ok in my CPP projects that use PCHs, it does not help at all in my C projects that do not use PCHs. Go to Symbol, Syntax highlighting, etc, has all been broken in those projects since the 10x pre-processor was implemented.

**Comment by slynch8** [comment]
Can you get around this for now by setting Force Includes in the workspace settings? They act like PCH but should work with C.

**Comment by slynch8** [comment]
You can also add defines to the workspace settings

**Comment by Muzza** [comment]
I tried adding to the Force Includes. I had to delete all the parsing info and restart 10x for it to apply, and it worked, but it also had a side effect that meant its not usable related to include guards.

```cpp
// ForceIncludeMe.h
#include "Header.h"
#include "OtherThings.h"

// Header.h
#ifndef __HEADER_H_
#define __HEADER_H_
// ALL code here is dulled out by the syntax highlighter because __HEADER_H_ is defined
#endif
```

**Comment by slynch8** [comment]
I don't follow, is this with your original example? The way I understand it, you have header2.h that doesn't include header.h so the parser doesn't know about the define. So you add header.h as a force include. This is equivalent to adding #include "header.h" to the top of header2.h, why would this cause a problem? or does header2.h use the same include guard as header.h?

The reason I want to check this works is if I find a way to do this automatically I'll be doing exactly the same thing.

![image](https://user-images.githubusercontent.com/7141856/154435360-ce266b51-e679-4dd1-a09a-ebe79bbcfa7a.png)

**Comment by Muzza** [comment]
Parser evaluates ForceIncludeMe.h which evaluates #include "Header.h" where it finds the #define __HEADER_H_ line.
Then when the parser is looking at Header.h to do syntax highlighting on it, it thinks __HEADER_H_ is defined, and so dulls out everything after #ifndef __HEADER_H_
At least that is what I assume is happening.

**Comment by slynch8** [comment]
oh I see, so everything is defined out in header.h. I just need to make sure I don't force include header.h in header.h

**Comment by slynch8** [comment]
I've stopped headers from including themselves in the latest build (0.31.4) which should fix the defined out problem you were seeing.
